// @form-create/core/src/frame/language.js
export default  {
    en: {
        required: '{title} is required',
        validate: {
            url: '{title} is not a valid url',
            email: '{title} is not a valid email',
            ip: '{title} {title} is not a valid ip',
            phone: '{title} {title} is not a valid phone',
            pattern: '{title} does not match pattern {pattern}',
            uppercase: '{title} must be all uppercase',
            lowercase: '{title} must be all lowercased',
            positive: '{title} is not a positive number',
            negative: '{title} is not a negative number',
            equal: '{title} is not equal to {equal}',
            min: '{title} cannot be less than {min}',
            max: '{title} cannot be greater than {max}',
            enum: '{title} must be one of {enum}',
            hasKeys: '{title} does not contain required fields {hasKeys}',
            minLen: '{title} must be at least {minLen}',
            maxLen: '{title} cannot be longer than {maxLen}',
            len: '{title} must be exactly {len}',
            integer: '{title} is not an integer',
            number: '{title} is not an number',
        }
    },
    'zh-cn': {
        required: '{title}不能为空',
        validate: {
            url: '{title}不是有效的 url 地址',
            email: '{title}不是有效的邮箱地址',
            ip: '{title}不是有效的 IP 地址',
            phone: '{title}不是正确的手机号',
            pattern: '{title}必须匹配 {pattern}',
            uppercase: '{title}必须全大写',
            lowercase: '{title}必须全小写',
            positive: '{title}不是正数',
            negative: '{title}不是负数',
            equal: '{title}必须等于 {equal}',
            min: '{title}必须大于等于 {min}',
            max: '{title}必须小于等于 {max}',
            enum: '{title}必须是 {enum} 之一',
            hasKeys: '{title}必须包含 {hasKeys} 字段',
            minLen: '{title}长度必须大于 {minLen}',
            maxLen: '{title}长度必须小于 {maxLen}',
            len: '{title}长度必须为 {len}',
            integer: '{title}必须为整数',
            number: '{title}必须为数字',
        },
    }
}
